docs(platform-wallet): error-code registry for the FFI result space (#4261)
dashevo/platform: packages/rs-platform-wallet-ffi/ERROR_CODE_REGISTRY.md

# FFI Error-Code Registry

Single source of truth for the integer values of
`PlatformWalletFFIResultCode` (`packages/rs-platform-wallet-ffi/src/error.rs`).

Every value in that enum is **public ABI**. `cbindgen` emits it into the
generated C header, and hosts compare against the integer — Swift
(`packages/swift-sdk/Sources/SwiftDashSDK/PlatformWallet/PlatformWalletResult.swift`)
mirrors it as a `RawRepresentable` enum, Kotlin
(`packages/kotlin-sdk/.../errors/DashSdkError.kt`) branches on it in
`fromPlatformWalletNative`. A shipped host binary that was compiled against
one numbering keeps using that numbering.

This file exists because several feature branches allocate into the same
integer range in parallel. A duplicate discriminant in two branches does **not**
produce a textual merge conflict — the second merge silently misclassifies
errors on every host — so allocations have to be reconciled here, in one place,

> rather than in each branch's diff.

Reviewer: <!-- thepastaclaw-review v1 finding=6aebeaeb0fd9 dedupe=0222ac8fab210938 -->
**💬 Nitpick: Duplicate Rust discriminants do not silently reach runtime**

A combined Rust enum containing two variants with the same explicit value fails compilation with E0081, and Swift rejects duplicate raw enum values as well. A textual merge can therefore leave an invalid tree, but that tree cannot successfully build and silently misclassify runtime errors. The actual ABI hazard is that independent branches or releases can reuse or renumber an integer relative to hosts already compiled against another meaning, as #3968 currently does with shipped code 26. State that cross-branch and cross-version failure mode instead.

```suggestion
This file exists because several feature branches allocate into the same
integer range in parallel. A combined Rust enum with duplicate discriminants
fails to compile, but independent branches can still reuse or renumber an ABI
value relative to a host that has already compiled the other meaning. Such a
version mismatch silently misclassifies errors on that host, so allocations
have to be reconciled here, in one place, rather than in each branch's diff.
```

<sub>source: ['codex']</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

You are right, and the file contradicted itself on it: the preamble said a duplicate discriminant "silently misclassifies errors on every host", while the code-32 section said the opposite — *"this one was not a paper conflict… produced a hard `error[E0081]: discriminant value 32 assigned more than once`"*.

The preamble now splits the two shapes explicitly:

- **Two different variant names on one integer** — the merged enum fails to compile with E0081. Loud, but only after someone merges both branches into one tree; neither branch's own CI sees it, because neither branch contains both variants. This is how the code-32 collision was caught.
- **The same meaning moving to a different integer, or a host mirror left un-updated** — nothing fails to compile, and a shipped host silently reads the new integer as whatever the old one meant. This is the failure the file mainly exists to prevent.

The collision-history paragraph that repeated the old claim was corrected the same way: it now says neither compiler ever sees the E0081 because neither tree contains both variants, rather than implying duplicates are inherently silent.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `690d464` — _Duplicate Rust discriminants do not silently reach runtime_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

## Rules

1. **Claim the next free integer** from the table below — the first value not
   listed as merged, proposed, or reserved. Do not reuse a gap unless this file
   marks it free.
2. **Record the claim in this file in the same PR** that adds the variant. A PR
   that adds a code without a row here is incomplete.
3. **Never renumber a code after it has shipped in a release.** Deprecate
   instead: leave the row, mark it deprecated, and allocate a new integer. Codes
   that are still only proposed (unmerged) may be renumbered to resolve a
   collision; codes on `v4.2-dev` may not.
4. **Do not reuse a retired integer.** Mark it reserved and move on.
5. **Update the mirrors in the same PR**: the Rust enum, the Swift
   `PlatformWalletResultCode` + its `init(result:)` switch, and — where the code
   deserves typed handling — the Kotlin `fromPlatformWalletNative` mapping and
   `DashSdkErrorTest`. Kotlin is allowed to be non-exhaustive: unmapped codes
   fall through to `PlatformWallet.Generic(code, …)`, which preserves the
   integer. Swift is exhaustive; an unmirrored code surfaces as

>    `.errorUnknown` there and loses its identity.

Reviewer: <!-- thepastaclaw-review v1 finding=237fa2d292c1 dedupe=b7cc539f42a1b759 -->
**🟡 Suggestion: Require the actual C-to-Swift result-code switch**

The generated C enum is first converted by `PlatformWalletResultCode.init(ffi:)` at `PlatformWalletResult.swift:75-138`. That switch has a default that maps an omitted native constant to `.errorUnknown`. The registry instead names a nonexistent `PlatformWalletResultCode.init(result:)`; the actual `init(result:)` belongs to the downstream `PlatformWalletError` conversion. A contributor could therefore add the Swift raw case and downstream error handling but omit `init(ffi:)`, losing the native code's identity before typed handling sees it. Require all three Swift locations explicitly.

```suggestion
5. **Update the mirrors in the same PR**: the Rust enum; the Swift
   `PlatformWalletResultCode`, its `init(ffi:)` switch that maps the generated C
   constants, and `PlatformWalletError` + its `init(result:)` switch; and — where
   the code deserves typed handling — the Kotlin `fromPlatformWalletNative`
   mapping and `DashSdkErrorTest`. Kotlin is allowed to be non-exhaustive:
   unmapped codes fall through to `PlatformWallet.Generic(code, …)`, which
   preserves the integer. Swift's `init(ffi:)` has an unknown-value fallback;
   omitting its mapping surfaces the new code as `.errorUnknown` and loses its
   identity.
```

<sub>source: ['codex']</sub>

Author: Fixed in 78023422b5e9c4788db08a1e06abffa29d467c43.

You're right that rule 5 named a `PlatformWalletResultCode.init(result:)` that does not exist — `init(result:)` belongs to the downstream `PlatformWalletError`. The gap you describe is real and #4204 is sitting in it right now: at `d78b940a03` it has the raw case and the `init(ffi:)` arm but no `PlatformWalletError` case, so the exhaustive `init(result:)` no longer compiles.

Rather than list the three locations, rule 5 now enumerates them **and states how each one fails**, since the two failure modes are opposite and that is the part worth remembering:

1. `PlatformWalletResultCode` — the raw case.
2. `PlatformWalletResultCode.init(ffi:)` — the arm mapping the generated C constant. Has a `default:` yielding `.errorUnknown`, so omitting it **compiles fine** and silently loses the code's identity before typed handling sees it.
3. `PlatformWalletError` — the typed case **and** its `init(result:)` arm. That switch is exhaustive with no `default:`, so adding (1) without this makes it non-exhaustive and the Swift package **stops compiling**.

Kotlin's non-exhaustive `Generic(code, …)` fallback is unchanged and now sits after the three Swift sites rather than being interleaved with them.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `690d464` — _Require the actual C-to-Swift result-code switch_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

6. **Blocks 98–99 are terminal sentinels** (`NotFound`, `ErrorUnknown`) and are
   not an allocation frontier. New codes go after the highest allocated value
   below them.

## Merged allocations (`v4.2-dev`)

These are shipped ABI. Do not renumber.

| Code | Name | Notes |
| ---: | --- | --- |
| 0 | `Success` | |
| 1 | `ErrorInvalidHandle` | |
| 2 | `ErrorInvalidParameter` | |
| 3 | `ErrorNullPointer` | |
| 4 | `ErrorSerialization` | |
| 5 | `ErrorDeserialization` | |
| 6 | `ErrorWalletOperation` | |
| 7 | `ErrorIdentityNotFound` | |
| 8 | `ErrorContactNotFound` | |
| 9 | `ErrorInvalidNetwork` | |
| 10 | `ErrorInvalidIdentifier` | |
| 11 | `ErrorMemoryAllocation` | |
| 12 | `ErrorUtf8Conversion` | |

> | 13 | `ErrorArithmeticOverflow` | Reserved slot — declared, no in-tree producer; holds the number for the mapping arriving via #3549 |

Reviewer: <!-- thepastaclaw-review v1 finding=9f9aa0523f7e dedupe=53f7e6a7e0ed1820 -->
**💬 Nitpick: ErrorArithmeticOverflow already has an in-tree producer**

The exact PR base already produces this code in `packages/rs-platform-wallet-ffi/src/shielded_send.rs:204-213`: `platform_wallet_shielded_estimate_fee` maps shielded fee-formula failures to `ErrorArithmeticOverflow`. That producer entered through commit `c9f8ef57925` and is present on `v4.1-dev`; the current #3549 diff does not add this mapping. Calling the slot producerless and awaiting #3549 gives incorrect provenance for an ABI registry.

```suggestion
| 13 | `ErrorArithmeticOverflow` | Produced by `platform_wallet_shielded_estimate_fee` when shielded fee computation overflows |
```

<sub>source: ['codex']</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

Confirmed on `v4.2-dev` at `5d68612a45`: `packages/rs-platform-wallet-ffi/src/shielded_send.rs:210` returns `PlatformWalletFFIResultCode::ErrorArithmeticOverflow`, with the contract described at line 180. So the row's "declared, no in-tree producer" was wrong, and crediting #3549 with the eventual mapping was misleading.

The row now names `shielded_send.rs` as the producer. It also records that the variant's own rustdoc in `error.rs` still calls itself a reserved slot with no producer — that comment is stale for the same reason, and should be corrected by whichever PR touches it next. Left as a note rather than a code change since this PR is documentation-only.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `690d464` — _ErrorArithmeticOverflow already has an in-tree producer_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

| 98 | `NotFound` | Sentinel — `Option` returned as an error |
| 99 | `ErrorUnknown` | Sentinel — unmapped/flattened errors |

**Next free integer: 34** (see the proposed table; 27–33 are claimed).

## Proposed allocations (open PRs)

Not yet ABI. Numbers here may still move; they move by agreement recorded in
this file.

| Code | Name | Owning PR | Status |
| ---: | --- | --- | --- |
| 27 | `ErrorStaleReservationToken` | #4185 | In review (also carried by #4256) |
| 28 | `ErrorReservationTokenConsumed` | #4185 | In review (also carried by #4256) |
| 29 | `ErrorReservationWalletMismatch` | #4185 | **Collision** — see below |
| 29 | `ErrorAssetLockInsufficientFunds` | #4184 | **Collision** — see below |
| 30 | — | — | **Unallocated.** Reserved in sibling comments only; see below |
| 31 | `ErrorSigningKeyUnavailable` | #4183 | In review (also carried by #4204) |
| 32 | `ErrorTransactionBuild` | #4247 | In review (also carried by #4256) |
| 33 | `ErrorTransactionSigning` | #4256 | In review |

Open PRs that touch `rs-platform-wallet-ffi` but claim **no** new code: #4186,

> #4191, #4194, #4195, #4240, #4251, #4258.

Reviewer: <!-- thepastaclaw-review v1 finding=176efb58c345 dedupe=45983cf92874650b -->
**🔴 Blocking: Registry omits incompatible allocations from active PRs**

The open-PR inventory and survey provenance omit branches that modify this exact ABI enum. At current head `5931df745a`, #3968 assigns `ErrorPersisterTransient = 26`, `ErrorPersisterFatal = 27`, and renumbers the already-shipped `ErrorTransactionBroadcastRejected` from 26 to 28. At head `93d0bd49b`, #3954 assigns `ErrorShutdownIncomplete = 27`. This conflicts with shipped code 26 and with #4185's code-27 claim, while #3968 and #3954 also assign different meanings to 27. An existing Swift host would classify #3968's persister-transient code 26 as `.errorTransactionBroadcastRejected`, and its actual rejection code 28 would fall through `init(ffi:)` to `.errorUnknown`. #4259 at `64146a2bb6` should also be recorded as carrying the same code-31 allocation as #4183. The supposedly complete no-new-code inventory additionally omits open PRs #3417, #3549, #3992, and #4243. Reconcile the incompatible claims and recompute the next-free value before presenting this file as the source of truth.

<sub>source: ['codex']</sub>

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `690d464` — _Registry omits incompatible allocations from active PRs_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610 — the last outstanding part of this finding.

Re-checked each sub-claim against `2d2c6c8` rather than trusting the earlier auto-resolve:

- **#3968's 26/27/28** — already recorded in the non-conforming table (rows for `ErrorPersisterTransient = 26`, `ErrorPersisterFatal = 27`, and the shipped `ErrorTransactionBroadcastRejected` renumbered 26 → 28). Landed earlier; still correct at head `5931df745a`.
- **#3954's `ErrorShutdownIncomplete = 27`** — recorded, and resolved the other way: #3954 was closed and its work merged as #4268, which is why 27 is in the merged table.
- **#4259's carried code 31** — recorded in the inherited-code table. Head refreshed to `9336bdbb71`.
- **The no-new-code inventory omitting #3417, #3549, #3992 and #4243** — this part was **not** fixed. It is now.

The inventory was rebuilt on 2026-08-03 from each PR's actual file list plus the `error.rs` at its head, and it now reads `#3417`, `#3549`, `#3992`, `#4186`, `#4191`, `#4194`, `#4195`, `#4243`. Four entries were also removed, each recorded with its reason so they are not silently re-added: #4240 and #4251 touch no file under this crate at their heads; #4258 merged into `v4.2-dev` on 2026-08-03; #4264 is closed, with its `error.rs` change carried by #4243. #4243 is called out explicitly — it does modify `error.rs`, but only to map new wallet errors onto the existing `ErrorInvalidParameter`, and this list tracks integer claims, not file touches.

## Contested and pending

### 29 — `ErrorReservationWalletMismatch` (#4185) vs `ErrorAssetLockInsufficientFunds` (#4184)

Both PR heads define code 29. This is the known collision: review on #4185
directed that PR to keep #4184's `29 = ErrorAssetLockInsufficientFunds` and move
`ErrorReservationWalletMismatch` to 30. That renumber has not landed on #4185's
head, and #4256 (stacked downstream) carries the pre-renumber `29`.

Resolution of record: **#4184 keeps 29; #4185 moves to 30**, propagated through
the Rust enum, the FFI `From` mapping, Swift `PlatformWalletResult`, Kotlin
`DashSdkError` (+ `DashSdkErrorTest`), and the JNI rustdoc — plus #4256, which
inherits the value.

### 30 — reserved in comments for a variant that no longer exists

`ErrorAssetLockCrossDomainConsentRequired` is named as the holder of 30 in
in-tree comments on #4183, #4204, and #4247/#4256's numbering rationale. It is
**not defined anywhere** — #4184, the PR that would have introduced it, does not
contain it after a re-scope. 30 is therefore free, and is the slot the #4185

> renumber above should take. The stale "reserved for the consent code" comments

Reviewer: <!-- thepastaclaw-review v1 finding=1a190de32e99 dedupe=00c53d965c81659e -->
**🔴 Blocking: Code 30 is marked free even though the registry assigns it to #4185**

The table marks 30 as `Unallocated`, and the later section explicitly calls it free, while the resolution of record assigns that value to #4185's `ErrorReservationWalletMismatch`; the stated next-free value of 34 also assumes that ownership. Rule 1 permits contributors to claim a gap when this file marks it free, so another PR could legitimately take 30 while #4185 follows the recorded renumber directive. Record 30 as allocated to #4185 and describe its current use of 29 only as stale surveyed-head state.

<sub>source: ['codex']</sub>

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `690d464` — _Code 30 is marked free even though the registry assigns it to #4185_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

The contradiction was real and still live at `2d2c6c8`, despite the earlier auto-resolve on this thread. The table rows and the frontier paragraph said `RESERVED`, but two other places said the opposite:

- the code-30 section: *"30 is free once more, and is deliberately not being reissued"*
- the collision history: *"**28 and 30 stay free.** Do not reissue them in this review cycle"*

Under rule 1 ("Do not reuse a gap unless this file marks it free") those two sentences licensed exactly the claim the rest of the file forbids, so the document carried two allocation frontiers at once.

All four places now say reserved-not-free and say it the same way. The code-30 heading is now "30 — vacated, then RESERVED (not free)", the body distinguishes *vacated* from *freed*, and the collision-history paragraph ends with an explicit pointer back to rule 1: reusing a gap requires this file to mark it free, and this file marks neither 28 nor 30 free. The frontier paragraph makes the same statement rather than just asserting 38.

**Next free integer: 34** — 27–33 are all claimed in the proposed table below.
(Before the 29/30 resolution landed this line disagreed with the table, which
still showed 30 as unallocated; 30 is now allocated to #4185 and the two agree.)

## Proposed allocations (open PRs)

Not yet ABI. Numbers here may still move; they move by agreement recorded in
this file.

| Code | Name | Owning PR | Status |
| ---: | --- | --- | --- |
| 27 | `ErrorStaleReservationToken` | #4185 | In review (also carried by #4256) |
| 28 | `ErrorReservationTokenConsumed` | #4185 | In review (also carried by #4256) |
| 29 | `ErrorAssetLockInsufficientFunds` | #4184 | In review — **keeps 29** (collision resolved) |
| 30 | `ErrorReservationWalletMismatch` | #4185 | In review — **moved 29 → 30** (collision resolved; #4256 has inherited it, #4196 inherits on restack) |
| 31 | `ErrorSigningKeyUnavailable` | #4183 | In review (also carried by #4204) |
| 32 | `ErrorTransactionBuild` | #4247 | In review (also carried by #4256) |
| 33 | `ErrorTransactionSigning` | #4256 | In review |

Open PRs that touch `rs-platform-wallet-ffi` but claim **no** new code: #4186,

> #4191, #4194, #4195, #4240, #4251, #4258.

Reviewer: <!-- thepastaclaw-review v1 finding=a8c90673c0f8 dedupe=027d3835fe2e3835 -->
**🟡 Suggestion: Make the no-new-code PR inventory complete and accurate**

The list does not match the open PRs that modify this crate. GitHub's exact PR file lists show that #3417, #3549, #3992, #4243, and #4264 modify `packages/rs-platform-wallet-ffi` without adding a result-code integer, but they are omitted. #4243 and #4264 even modify `error.rs`, intentionally mapping new wallet errors to the existing `ErrorInvalidParameter` code. Conversely, the exact surveyed heads for listed PRs #4240 and #4251 contain no file under this crate. These discrepancies do not affect the next-free integer, but they make the registry's open-PR survey incomplete and inaccurate.

```suggestion
Open PRs that touch `rs-platform-wallet-ffi` but claim **no** new code: #3417,
#3549, #3992, #4186, #4191, #4194, #4195, #4243, #4258, #4264.
```

<sub>source: ['claude', 'codex']</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

Rebuilt the inventory on 2026-08-03 from each PR's file list and the `error.rs` at its head rather than applying the suggestion verbatim, because two of its entries had moved since it was written:

- **`#4264` is closed.** Its `error.rs` change — mapping new wallet errors onto the existing `ErrorInvalidParameter` — is carried by `#4243`, which is open and is in the list.
- **`#4258` has merged** into `v4.2-dev` (`ce8233edb7`). It claimed no code, so the merged table is unchanged, but it is no longer an open PR.

The list is now `#3417`, `#3549`, `#3992`, `#4186`, `#4191`, `#4194`, `#4195`, `#4243`. Your two removals are applied as written: #4240 and #4251 touch no file under this crate at their heads. All four removals are recorded in a short table with the reason for each, so a later pass does not re-add them from an older revision of this file. #4243's `error.rs` involvement is called out explicitly for the same reason — touching `error.rs` is not the same as claiming an integer.

Two more carry a code they did not allocate, inherited from the PR they are
stacked on rather than claimed fresh — they must not be read as a second claim
on the number:

| Code | Name | Carried by | Allocated to |
| ---: | --- | --- | --- |
| 31 | `ErrorSigningKeyUnavailable` | #4204, #4259 | #4183 |

#4196 also claims no new integer: it adds a token-less
`PlatformWalletError::StaleReservation` variant and deliberately routes it
through the **existing** `ErrorStaleReservationToken`, so it allocates nothing
and only has to follow that code's number (see below).

### Non-conforming allocations (rebase required)

These branches allocate into the same range from a stale base. They are listed
here rather than in the proposed table because their numbers cannot stand as
written — each row is a claim to be **withdrawn and reissued**, not an

> allocation of record.

Reviewer: <!-- thepastaclaw-review v1 finding=44fe1e6c5969 dedupe=906d592abfe7b0b5 -->
**🟡 Suggestion: Do not classify #3968 as a stale-base rebase**

The surveyed #3968 head `5931df745a` already contains the registry's exact base `ed4116b26c`: `git merge-base` returns `ed4116b26c`, and merge commit `debf67bdae` brought that base into the branch. Its diff from that base explicitly adds persister codes 26 and 27 and moves `ErrorTransactionBroadcastRejected` from 26 to 28. The section heading, stale-base explanation, and instruction at lines 224-225 therefore prescribe a rebase that has already occurred. The substantive remedy stated elsewhere is correct: #3968 must edit the enum to restore broadcast rejection to 26 and assign fresh values to both persister codes.

<sub>source: ['codex']</sub>

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `e915e5c` — _Do not classify #3968 as a stale-base rebase_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

Author: Addressed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610, with a correction to the premise.

Your merge-base check is right and I reproduced it: `git merge-base 5931df745a ed4116b26c` returns `ed4116b26c`, so #3968's head does contain the base this file was originally compiled against. Calling it a stale-base branch on that basis was wrong.

But the base has since moved. `git merge-base --is-ancestor 5d68612a45 5931df745a` fails — #3968 does **not** contain the current `v4.2-dev` head, which is where #4268's merged `ErrorShutdownIncomplete = 27` lives. So a rebase is genuinely outstanding; it just is not the remedy.

The section heading is now "Non-conforming allocations (withdraw and reissue)" instead of "(rebase required)", and it opens with exactly that distinction: the rebase is a precondition, #3968 is behind the current base rather than un-rebased, and rebasing resolves nothing on its own because git sees no conflict in any of it — the branch has to edit its own enum. Which is the substantive remedy you identified, and it is unchanged.

JNI rustdoc, the Swift raw value, and Kotlin's `fromPlatformWalletNative` branch,
class KDoc, `WalletManagerNative` KDoc and `DashSdkErrorTest` offset assertion.
#4256's other codes are untouched: it keeps 32 (shared with #4247) and 33.

Note that neither #4184 nor #4256 was ever blocked by CI on this. Both are
MERGEABLE with green checks, because two branches assigning the same
discriminant produce no textual conflict — the collision surfaces only as an
E0081 after a textual merge, or silently as a wrong error code on the host.
That is the whole reason this file exists.

**Still outstanding:** #4196 (see below).

### 30 — allocated to #4185; the old "consent code" reservation was stale

`ErrorAssetLockCrossDomainConsentRequired` is named as the holder of 30 in
in-tree comments on #4183, #4204, and #4247/#4256's numbering rationale. It is
**not defined anywhere** — #4184, the PR that would have introduced it, does not
contain it after a re-scope.

Verified 2026-08-01 by reading `packages/rs-platform-wallet-ffi/src/error.rs` at
the head of **every one of the 62 open PRs**: no PR anywhere defines a code 30.

> 30 was therefore genuinely free, and #4185 has taken it. The stale "reserved for

Reviewer: <!-- thepastaclaw-review v1 finding=0d0f830d9500 dedupe=85fcec3933e2a493 -->
**💬 Nitpick: Qualify the code-30 sweep as pre-allocation**

The exact surveyed heads later cited for #4185 (`6c37e8679e`) and #4256 (`9481e5783b`) both define `ErrorReservationWalletMismatch = 30`, so the unqualified statement that no open PR defines code 30 is false for the recorded survey. The cited #4247 head `0dcdc743e7` also does not mention `ErrorAssetLockCrossDomainConsentRequired`; the stale reservation appears on #4183, #4204, and #4256's pre-renumber rationale. The intended conclusion remains valid—no unrelated PR competes with #4185 for 30—but the provenance needs to distinguish the stale historical reservation from the post-allocation surveyed state.

```suggestion
`ErrorAssetLockCrossDomainConsentRequired` is named as the holder of 30 in
in-tree comments on #4183 and #4204, and in #4256's pre-renumber numbering
rationale. It is **not defined anywhere** — #4184, the PR that would have
introduced it, does not contain it after a re-scope.

At the surveyed heads, #4185 and its downstream #4256 define
`ErrorReservationWalletMismatch = 30`; no unrelated open PR defines code 30.
The stale consent-code reservation therefore does not conflict with #4185's
allocation.
```

<sub>source: ['codex']</sub>

Author: Fixed in 78023422b5e9c4788db08a1e06abffa29d467c43.

Correct on both counts, and the first one is the kind of overstatement that undermines the rest of the file. "No PR anywhere defines a code 30" was false for the very heads the provenance cites — #4185 and its downstream #4256 both defined `ErrorReservationWalletMismatch = 30` at the surveyed heads. That was the allocation, not a competing claim, but the sentence did not say so.

It now reads: **no PR unrelated to #4185 defines a code 30** — which is the claim that actually supports the conclusion, stated as a qualification rather than left implicit. The sweep result is unchanged: nothing contested 30, #4185's claim stood, and the stale consent-code reservation never conflicted with it.

The list of branches carrying that stale reservation is corrected to **#4183 and #4204, plus #4256's pre-renumber rationale**. #4247 was never one of them.

One update since you wrote this: #4183's stale comment is gone as of its 2026-08-03 rebase, so #4204 is now the only branch still carrying one. That is recorded in the same section.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `e915e5c` — _Qualify the code-30 sweep as pre-allocation_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

  #4204. No conflict; recorded so the number is not double-counted.

Both #3968 and #3954 need a rebase onto current `v4.2-dev` and fresh integers
from the frontier (34+); #4185's 27/28 are the older claim and should stand.

### 26 — `ErrorStaleReservationToken` on #4196 collides with merged ABI

#4196 (stacked on #4185) branched before `26 = ErrorTransactionBroadcastRejected`
merged, and its head numbers the reservation trio **26 / 27 / 28**. Merging it as
it stands would give 26 two meanings and would contradict #4185's own
27 / 28 / 30 for the same three names. #4196 needs a rebase and must adopt
whatever numbering #4185 lands with. No new integers are needed for it.

**All three of those numbers come from the copy of #4185 that #4196 carries, not
from #4196's own commits.** Restacking onto #4185's head therefore fixes the
trio for free — including `ErrorReservationWalletMismatch` 28 → 30, which #4196
never had to move itself. The one number #4196 does own is a doc reference: its
`StaleReservation` variant and the matching Kotlin KDoc both cite
`ErrorStaleReservationToken` as **26**, and that becomes **27** post-restack.

> So the number #4196 must chase is 27, not 30.

Reviewer: <!-- thepastaclaw-review v1 finding=291a45e56c14 dedupe=1fb2a53c74fb9af2 -->
**🟡 Suggestion: List every numeric reference owned by #4196**

Comparing #4196's three own commits using `911c8f7264..ea4f783490` shows three changed or added references to native code 26 that must become 27 during the restack: the mapping-test rustdoc in `rs-platform-wallet-ffi/src/error.rs`, the rewritten `StaleReservationToken` KDoc in `DashSdkError.kt`, and the public V2 broadcast KDoc in `ManagedCoreWallet.kt`. Conversely, the `PlatformWalletError::StaleReservation` variant only refers to the FFI code symbolically and contains no number. The current two-reference description can leave public Kotlin boundary documentation advertising code 26 after the runtime mapping moves to 27.

```suggestion
**All three reservation-trio values come from the copy of #4185 that #4196
carries, not from #4196's own commits.** Restacking onto #4185's head therefore
fixes the trio for free — including `ErrorReservationWalletMismatch` 28 → 30.

The numeric references in #4196's own changed hunks still need explicit updates:
the mapping-test rustdoc in `rs-platform-wallet-ffi/src/error.rs`, the rewritten
`StaleReservationToken` KDoc in `DashSdkError.kt`, and the public V2 broadcast
KDoc in `ManagedCoreWallet.kt` all cite **26** and must become **27**. The
`PlatformWalletError::StaleReservation` variant itself contains no numeric
reference. Thus #4196 allocates nothing, but it must update all three documents
to code 27 during the restack.
```

<sub>source: ['codex']</sub>

Author: Addressed in 78023422b5e9c4788db08a1e06abffa29d467c43, though the finding has been overtaken by events in the best way.

**#4196 has restacked.** Its head is now `12492e8c54`, not `ea4f783490`. Verified: `ErrorStaleReservationToken = 34`, `ErrorReservationTokenConsumed = 35`, `ErrorReservationWalletMismatch = 36`, `ErrorShutdownIncomplete = 27` present from the merged base, #4185's head `8813e98533` is an ancestor, and the PR is MERGEABLE against `v4.2-dev`.

So the target number is 34, not the 27 the suggestion assumed — that was written before the trio's third move — and the references you identified were carried along with the restack rather than needing separate updates. Confirmed at `12492e8c54`:

- `DashSdkError.kt` — `StaleReservationToken` KDoc reads "native code 34"; `fromPlatformWalletNative` maps `34 -> PlatformWallet.StaleReservationToken`.
- `ManagedCoreWallet.kt` — V2 broadcast KDoc reads "native code 34, shared with the deferred-token surface"; the remaining mentions are symbolic `[StaleReservationToken]` links with no number.
- `PlatformWalletError::StaleReservation` — symbolic only, no number, exactly as you noted.

Your substantive point stands and is recorded: the file previously described **two** references when #4196 owned three (including the `ManagedCoreWallet.kt` public boundary KDoc), and named the variant as one of them when it carries no number. That inventory is corrected, and the section is now a resolution rather than an open item. The account of why the restack was hard — #4185's `registered_height`, `WalletRemoved`, and the in-place rewrite of `RESERVATION_MAX_AGE_BLOCKS` — is kept, since that was the substance of the delay.

the reservation trio moved to 34–36, but they are deliberately left unclaimed
rather than back-filled, so that the trio stays contiguous and no number is
reused within a single review cycle. A new code should take 37 unless it has a
reason to sit next to something.

## Proposed allocations (open PRs)

Not yet ABI. Numbers here may still move; they move by agreement recorded in
this file.

| Code | Name | Owning PR | Status |
| ---: | --- | --- | --- |
| 28 | *(free)* | — | Vacated by #4185/#4256 on 2026-08-02; not reissued |
| 29 | `ErrorAssetLockInsufficientFunds` | #4184 | In review — **keeps 29** (collision resolved) |
| 30 | *(free)* | — | Vacated by #4185/#4256 on 2026-08-02; not reissued |
| 31 | `ErrorSigningKeyUnavailable` | #4183 | In review (also carried by #4204, #4259) |
| 32 | `ErrorTransactionBuild` | #4247 | In review (also carried by #4256) |
| 33 | `ErrorTransactionSigning` | #4256 | In review |
| 34 | `ErrorStaleReservationToken` | #4185 | In review — **moved 27 → 34** (also carried by #4256; #4196 inherits on restack) |
| 35 | `ErrorReservationTokenConsumed` | #4185 | In review — **moved 28 → 35** (also carried by #4256) |

> | 36 | `ErrorReservationWalletMismatch` | #4185 | In review — **moved 30 → 36** (also carried by #4256) |

Reviewer: <!-- thepastaclaw-review v1 finding=bfb2d6e2c256 dedupe=e539bbad72e02898 -->
**🔴 Blocking: Reserve 28 and 30 before declaring 37 next**

Rule 1 directs contributors to claim the first value not merged, proposed, or reserved and explicitly permits reuse when the registry marks a gap free. This section and the table mark 28 and 30 as free, making 28 the next allocatable value under the authoritative rule, while simultaneously declaring 37 next and instructing readers not to reissue either gap during this review cycle. Mark 28 and 30 temporarily reserved so the registry gives one unambiguous allocation frontier.

```suggestion
**Next allocatable integer: 37** — 27 is merged; 29 and 31–36 are claimed
in the proposed table below. **28 and 30 are temporarily reserved**: #4185 and
#4256 vacated them when the reservation trio moved to 34–36, but they must not
be reissued during this review cycle. A new code must therefore take 37.

## Proposed allocations (open PRs)

Not yet ABI. Numbers here may still move; they move by agreement recorded in
this file.

| Code | Name | Owning PR | Status |
| ---: | --- | --- | --- |
| 28 | *(temporarily reserved)* | — | Vacated by #4185/#4256 on 2026-08-02; do not reissue during this review cycle |
| 29 | `ErrorAssetLockInsufficientFunds` | #4184 | In review — **keeps 29** (collision resolved) |
| 30 | *(temporarily reserved)* | — | Vacated by #4185/#4256 on 2026-08-02; do not reissue during this review cycle |
| 31 | `ErrorSigningKeyUnavailable` | #4183 | In review (also carried by #4204, #4259) |
| 32 | `ErrorTransactionBuild` | #4247 | In review (also carried by #4256) |
| 33 | `ErrorTransactionSigning` | #4256 | In review |
| 34 | `ErrorStaleReservationToken` | #4185 | In review — **moved 27 → 34** (also carried by #4256; #4196 inherits on restack) |
| 35 | `ErrorReservationTokenConsumed` | #4185 | In review — **moved 28 → 35** (also carried by #4256) |
| 36 | `ErrorReservationWalletMismatch` | #4185 | In review — **moved 30 → 36** (also carried by #4256) |
```

<sub>source: ['codex']</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

The contradiction was real and still live at `2d2c6c8`, despite the earlier auto-resolve on this thread. The table rows and the frontier paragraph said `RESERVED`, but two other places said the opposite:

- the code-30 section: *"30 is free once more, and is deliberately not being reissued"*
- the collision history: *"**28 and 30 stay free.** Do not reissue them in this review cycle"*

Under rule 1 ("Do not reuse a gap unless this file marks it free") those two sentences licensed exactly the claim the rest of the file forbids, so the document carried two allocation frontiers at once.

All four places now say reserved-not-free and say it the same way. The code-30 heading is now "30 — vacated, then RESERVED (not free)", the body distinguishes *vacated* from *freed*, and the collision-history paragraph ends with an explicit pointer back to rule 1: reusing a gap requires this file to mark it free, and this file marks neither 28 nor 30 free. The frontier paragraph makes the same statement rather than just asserting 38.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `c6fb09a` — _Reserve 28 and 30 before declaring 37 next_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

Open PRs that touch `rs-platform-wallet-ffi` but claim **no** new code: #4186,

> #4191, #4194, #4195, #4240, #4251, #4258.

Reviewer: <!-- thepastaclaw-review v1 finding=c97c91dc8fd8 dedupe=1618267ed94a179f -->
**💬 Nitpick: Fix the MD018 warnings for leading PR references**

Lines beginning directly with `#<number>` are interpreted by markdownlint as malformed ATX headings and trigger MD018. Wrap each leading PR reference in backticks or escape its hash. This affects lines 105, 115, 133, 137, 180, 185, 225, 231, 235, 247, 248, 283, 297, 359, 361, and 365–367.

<sub>source: ['coderabbit']</sub>

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `2d2c6c8` — _Fix the MD018 warnings for leading PR references_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

Flagging that the auto-resolve on this thread was premature: `markdownlint-cli2` still reported **18** MD018 violations at `2d2c6c8`. The lines beginning with a bare `#<number>` were still there — 106, 116, 134, 138, 215, 220, 260, 266, 270, 282, 283, 318, 332, 394, 396, 400, 401, 402.

Fixed by rewording so the PR reference is no longer the first token on the line (`PR #3968 must keep 26…`, `the #4185 numbering of…`), rather than by escaping or code-fencing. That keeps GitHub's PR autolinks intact, which `\#4196` or `` `#4196` `` would have broken.

`markdownlint-cli2` now reports **0** MD018 on this file. MD004 also went to 0 in the same pass — the file had mixed `*`/`-` top-level bullets, now all `*`. MD013 is down from 19 to 18, all of them long table rows.

> #4191, #4194, #4195, #4240, #4251, #4258.

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

Flagging that the auto-resolve on this thread was premature: `markdownlint-cli2` still reported **18** MD018 violations at `2d2c6c8`. The lines beginning with a bare `#<number>` were still there — 106, 116, 134, 138, 215, 220, 260, 266, 270, 282, 283, 318, 332, 394, 396, 400, 401, 402.

Fixed by rewording so the PR reference is no longer the first token on the line (`PR #3968 must keep 26…`, `the #4185 numbering of…`), rather than by escaping or code-fencing. That keeps GitHub's PR autolinks intact, which `\#4196` or `` `#4196` `` would have broken.

`markdownlint-cli2` now reports **0** MD018 on this file. MD004 also went to 0 in the same pass — the file had mixed `*`/`-` top-level bullets, now all `*`. MD013 is down from 19 to 18, all of them long table rows.

written — each row is a claim to be **withdrawn and reissued**, not an
allocation of record.

| Code | Name | Owning PR | Conflict |
| ---: | --- | --- | --- |
| 26 | `ErrorPersisterTransient` | #3968 | Contradicts **merged ABI** — 26 is `ErrorTransactionBroadcastRejected` |
| 27 | `ErrorPersisterFatal` | #3968 | Contradicts **merged ABI** — 27 is #4268's `ErrorShutdownIncomplete` (was a #4185 collision until 2026-08-02) |
| 28 | `ErrorTransactionBroadcastRejected` | #3968 | **Renumbers a shipped code** 26 → 28 — forbidden by rule 3 |

#3954's `ErrorShutdownIncomplete = 27` used to sit in this table. It is gone
because that claim **won**: #3954 was closed and superseded by **#4268**, which
merged 27 into `v4.2-dev` on 2026-08-02. See the collision history below.

#3968 is the serious one: rule 3 forbids renumbering a code that has shipped,
and `ErrorTransactionBroadcastRejected = 26` is merged ABI. Moving it to 28
would silently reinterpret every 26 an already-compiled host returns. #3968 must
keep 26 where it is and take fresh integers from the frontier (37+) for its two
persister codes. Its 27 is now doubly wrong: 27 is merged ABI

> (`ErrorShutdownIncomplete`), so rule 3 protects it too.

Reviewer: <!-- thepastaclaw-review v1 finding=aab53643892a dedupe=fdf28ac4b17a1871 -->
**🔴 Blocking: Advance #3968's reissue frontier to 38**

Introduced by the latest delta. Code 37 is now allocated to #4204 and lines 80–85 correctly declare 38 as the next allocatable integer, but this paragraph still directs #3968 to obtain its two persister codes from `37+`. Lines 260–262 repeat the stale instruction. Following that branch-specific guidance would allow #3968 to claim 37 and recreate the collision with `ErrorShieldedInviteAlreadyClaimed`; both references must advance to `38+`.

<sub>source: ['codex']</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

Confirmed both references were stale. `#3968 must keep 26 where it is and take fresh integers from the frontier (37+)` in the non-conforming section, and the same `37+` in the 27/28 detail section, both now read **38+** — matching the frontier paragraph, which is unchanged at 38.

Two additions beyond the literal text of the finding, because "38+" alone would still have left an ambiguity for that branch: both places now also state that 28 and 30 are reserved rather than free, so #3968 cannot read the gaps as available. The free-vs-reserved contradiction that made that reading possible is fixed in the same commit (see the code-30 section and the collision history).

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `9f3dab1` — _Advance #3968's reissue frontier to 38_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

`ErrorReservationWalletMismatch = 32` produced a hard
`error[E0081]: discriminant value 32 assigned more than once`. Resolution of
record: **#4204 moves 32 → 37**, the frontier. `ErrorTransactionBuild` keeps 32.

The numbering was the lesser half of the defect. The code was **unmirrored on
both hosts** — absent from Swift's `PlatformWalletResultCode` and from Kotlin's
`fromPlatformWalletNative`. Per rule 5 that means Swift rendered it
`.errorUnknown` (identity lost), while Kotlin fell through to `Generic(32, …)`
— and in any tree carrying #4185's `ErrorReservationWalletMismatch = 32`,
Kotlin actively **misclassified** "shielded invite already claimed" as
"reservation wallet mismatch". That is the exact silently-wrong-error-on-every-host
failure this file's preamble describes, and it landed on the shielded-invite
claim-recovery path (the error is raised from four sites in
`wallet/shielded/operations.rs`, three of them inside the recovery function).

Fixed on #4204 together with the renumber: typed
`PlatformWallet.ShieldedInviteAlreadyClaimed` (terminal, `isRetryable = false`),
the Swift case and its `init(ffi:)` arm, and a `DashSdkErrorTest` assertion that

> pins 37 so a future move off the frontier fails the suite instead of the hosts.

Reviewer: <!-- thepastaclaw-review v1 finding=834e05ad7011 dedupe=cdd88c7b69b4b886 -->
**🟡 Suggestion: Do not mark #4204's Swift mirror fixed yet**

Introduced by the latest delta. At current #4204 head `d78b940a03`, Swift contains `errorShieldedInviteAlreadyClaimed = 37` and the corresponding `PlatformWalletResultCode.init(ffi:)` arm, but `PlatformWalletError` has no typed case and its exhaustive `init(result:)` switch has no `.errorShieldedInviteAlreadyClaimed` arm. The switch is therefore non-exhaustive after adding the raw case, so the Swift package cannot compile as written. Record the Swift work as incomplete until #4204 adds the downstream typed error conversion required by rule 5.

```suggestion
Partially fixed on #4204 together with the renumber: typed
`PlatformWallet.ShieldedInviteAlreadyClaimed` (terminal, `isRetryable = false`),
the Swift raw case and its `init(ffi:)` arm, and a `DashSdkErrorTest` assertion
that pins 37. The Swift `PlatformWalletError` case and its `init(result:)` arm
are still missing, so typed Swift error handling remains pending.
```

<sub>source: ['codex']</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

Verified against #4204 at `d78b940a03` before changing the text. `PlatformWalletResult.swift` there has `case errorShieldedInviteAlreadyClaimed = 37` (line 96) and the matching `init(ffi:)` arm (line 159), but `PlatformWalletError` has no `.shieldedInviteAlreadyClaimed` case, and its `init(result:)` switches over `PlatformWalletResultCode` with no `default:` — so adding the raw case makes that switch non-exhaustive and the Swift package does not build as written.

The section now says **partially** fixed, lists what actually landed (typed Kotlin case, the Swift raw case + `init(ffi:)` arm, the `DashSdkErrorTest` pin on 37) separately from what is missing, and states that rule 5's Swift clause is not yet satisfied. The proposed-table row for 37 is unchanged — the number is settled; it is the mirror that is incomplete.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `9f3dab1` — _Do not mark #4204's Swift mirror fixed yet_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

the same thing in both enums.

## Survey provenance

Compiled 2026-08-01 against `v4.2-dev` at `ed4116b26c`, and **re-verified
2026-08-02 against `v4.2-dev` at `5d68612a45`**, which is where
`ErrorShutdownIncomplete = 27` (#4268, `429667e723`) entered the merged table.
The 2026-08-02 pass re-read the added discriminants at the head of every open PR
that touches `error.rs`, `DashSdkError.kt` or `PlatformWalletResult.swift`
(#3968, #4183, #4184, #4185, #4186, #4191, #4194, #4195, #4196, #4204, #4243,
#4247, #4256, #4259) and confirmed the only claims in the 27–36 range are the
ones tabled above — in particular that 32 and 33 were **already taken** by
#4247/#4256, which is why the trio went to 34–36 rather than 32–34.

PR heads of record: #3954 `93d0bd49b7` (closed), #3968 `5931df745a`, #4183
`2cd948331b`, #4184 `bd19a3e020`, **#4185 `3dec774929`** (post-34/35/36 move),
#4186 `6f7abbadc1`, #4191 `8acb0bd14c`, #4194 `9efc0b7e3a`, #4195 `4f2eb06d64`,
#4196 `ea4f783490`, #4204 `7bc8a845c6`, #4240 `9328609a16`, #4247 `0dcdc743e7`,
#4251 `176f8ed3eb`, **#4256 `8febac177c`** (post-34/35/36 move), #4258
`5adfc40032`, #4259 `4270d827c2`. Rows describing open PRs reflect those heads

> and go stale as the PRs are updated; the merged table does not.

Reviewer: <!-- thepastaclaw-review v1 finding=dc9bafaf4c7d dedupe=ca4acbacfd122b68 -->
**🟡 Suggestion: Record the post-renumber #4204 head in provenance**

Introduced by the latest delta. The code-37 row and resolved-collision section describe #4204 state introduced by `d78b940a03`, but the provenance still names `7bc8a845c6` and says the rows reflect the listed heads. That recorded head contains no `ErrorShieldedInviteAlreadyClaimed` variant at all, so it cannot substantiate either the temporary code-32 allocation or the move to 37 and its host mappings. Add a 2026-08-03 revalidation and record `d78b940a03` as #4204's post-renumber head.

<sub>source: ['codex']</sub>

Author: Fixed in 9f3dab1753f700ac489c717e2dbb0a97ec34c610.

Rather than adding a single #4204 line, the whole provenance block was re-read from GitHub on 2026-08-03 and rewritten as a dated table, because several other heads had moved too (#4185, #4186, #4194, #4195, #4240, #4247, #4256 and #4259 were all stale, and #4258 had merged).

`#4204` is recorded at `d78b940a03` with the note "post-renumber 32 → 37; pre-#4268 base". The 2026-08-03 pass also confirms what the code-37 row and the resolved-collision section actually rest on: 32/33 are still #4247/#4256's, 37 is #4204's post-renumber code, nothing in flight has taken 28, 30 or 38, and #4247's head now carries #4185's 34/35/36 trio by inheritance rather than as a second claim.

The closing paragraph now states that every head in the table was read from GitHub on the date given rather than carried forward from a previous revision of this file — which is the mistake that produced the `#4185 0b0d5c76d6 (post-renumber)` entry the note already documents.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `9f3dab1` — _Record the post-renumber #4204 head in provenance_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

**Next allocatable integer: 38** — 27–37 are all claimed (27 merged; 29 and
31–37 in the proposed table below; 28 and 30 reserved). **28 and 30 are
RESERVED, not free**: #4185 and #4256 vacated them when the reservation trio
moved to 34–36, but they are deliberately left unclaimed rather than
back-filled, so that the trio stays contiguous and no number is reused within a
single review cycle. Rule 1's "do not reuse a gap unless this file marks it
free" applies — this file does **not** mark 28 or 30 free, so the frontier is
the only allocation source and a new code takes 38.

## Proposed allocations (open PRs)

Not yet ABI. Numbers here may still move; they move by agreement recorded in
this file.

| Code | Name | Owning PR | Status |
| ---: | --- | --- | --- |
| 28 | *(reserved — vacated)* | — | Vacated by #4185/#4256 on 2026-08-02; RESERVED, not reissuable — the next-free frontier is the only allocation source |
| 29 | `ErrorAssetLockInsufficientFunds` | #4184 | In review — **keeps 29** (collision resolved) |
| 30 | *(reserved — vacated)* | — | Vacated by #4185/#4256 on 2026-08-02; RESERVED, not reissuable — the next-free frontier is the only allocation source |

> | 31 | `ErrorSigningKeyUnavailable` | #4183 | In review (also carried by #4204, #4259) |

Reviewer: <!-- thepastaclaw-review v1 finding=5f1e4df7e508 dedupe=25b769aa9b659fdf -->
**🔴 Blocking: Record code 31 as merged ABI**

New current-base finding. Current v4.2-dev at `f53e5eef0a` contains `ErrorSigningKeyUnavailable = 31`, merged by #4183 in `189a3abb1c`, together with the Rust C-facing discriminant and complete Swift and Kotlin mirrors. This row still places 31 under “Proposed allocations,” whose introduction explicitly permits renumbering, contradicting rule 3 for a value now present on v4.2-dev. Move code 31 into the merged table and update the proposed, inherited-code, collision-history, open-PR, and provenance text that still treats #4183 as open.

<sub>source: ['codex']</sub>

Author: Fixed in `ae21604bae`.

The finding is correct, and the sweep it asked for turned up three more merged PRs the document was still treating as open.

**Verification base.** I did not verify against the `f53e5eef0a` cited in the finding, nor against the `5d68612a45` this file was last compiled against. The base has moved twice since; I fetched and verified against the current `origin/v4.2-dev` at **`97904ed2fc`**.

**Code 31 → merged table.** `packages/rs-platform-wallet-ffi/src/error.rs` at `97904ed2fc` runs 0–27 contiguously and then `ErrorSigningKeyUnavailable = 31`, with 28/29/30 and everything from 32 up absent. `git log -S` pins all three sites to one commit, `189a3abb1c` (#4183, stacked on #4191):

- Rust discriminant `ErrorSigningKeyUnavailable = 31`
- Swift `PlatformWalletResultCode.errorSigningKeyUnavailable = 31` (line 87), its `init(ffi:)` arm (150), the typed `PlatformWalletError` case (281) and its `init(result:)` arm (350)
- Kotlin `31 -> PlatformWallet.SigningKeyUnavailable` in `fromPlatformWalletNative` (`DashSdkError.kt:353`)

So rule 5 was satisfied in the merging commit, and rule 3 now protects the number. It was the only value in the proposed table that has merged; the frontier is unchanged at 38.

**PR states moved to merged** (all four merged 2026-08-04; each SHA below is the merge commit on `v4.2-dev`, confirmed an ancestor of `97904ed2fc` with `git merge-base --is-ancestor` — not a PR head SHA):

| PR | Merge commit | Claimed a code? |
| --- | --- | --- |
| #4191 | `0e2282b586` | No — only `dashpay.rs` in this crate |
| #4183 | `189a3abb1c` | **Yes — 31**, now in the merged table |
| #4277 | `6704a41a85` | No — modifies `error.rs`, but routes `TxMetadataPayloadTooLarge` onto the existing `ErrorInvalidParameter` |
| #4251 | `7afc8a8ff3` | No — touches no file under this crate |

`#4277` was not in the document at all; it is now recorded as the merged precedent for "touches `error.rs`, allocates nothing".

**Dependent sections updated,** so nothing left implies a merged code may still move: the frontier breakdown, the proposed table (plus an explicit note that 31 left it), the no-new-code open-PR inventory and its removed-entries table, the inherited-code table (31 is trunk now, not an inheritable claim — verified present at the heads of all eleven rebased branches: #4184, #4185, #4186, #4194, #4195, #4196, #4204, #4240, #4247, #4256, #4259), the collision-history bullet list, the `#3968` stale-base reference, the 31-vs-33 note (collapsing 31 is no longer available; only #4256's 33 is still an open decision), and the survey provenance with a refreshed heads-of-record table.

**Re-verified rather than carried forward,** since re-dating the provenance section would otherwise have preserved claims that are now false. At current heads: 29 is still #4184's, 32/33 still #4247/#4256's, 34/35/36 still the trio, 37 still #4204's, #3968 still numbers 26/27/28, and nothing has taken 28, 30 or 38. Two claims the document made are no longer true and were corrected: #4204's Swift typed-case gap is closed at `4efecd5b71` and #4256's missing Swift raw cases are closed at `862036b18d`. The stale `ErrorAssetLockCrossDomainConsentRequired` comments are gone from every branch that carried them (#4204 picked up the corrected text from trunk on rebase).

**One thing I did not change, flagged for your call.** Two pre-existing SHA references in the merged table and the removed-entries table label PR *head* commits as merge commits: #4268 is given as `429667e723` and #4258 as `ce8233edb7`, and neither is an ancestor of `v4.2-dev` (both PRs were squash-merged; the actual merge commits are `1e29d2f34c` and `5d68612a45`). It is cosmetic and outside this finding, and `5d68612a45` is load-bearing elsewhere in the document as the prior base, so I left both alone rather than untangle them in a fix for something else. Happy to correct them in a follow-up if you want the convention made uniform.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `ae21604` — _Record code 31 as merged ABI_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

failure this file's preamble describes, and it landed on the shielded-invite
claim-recovery path (the error is raised from four sites in
`wallet/shielded/operations.rs`, three of them inside the recovery function).

**Partially** fixed on #4204 together with the renumber. Landed at head
`d78b940a03`: the typed Kotlin
`PlatformWallet.ShieldedInviteAlreadyClaimed` (terminal, `isRetryable = false`),
the Swift `PlatformWalletResultCode.errorShieldedInviteAlreadyClaimed = 37` raw
case with its `init(ffi:)` arm, and a `DashSdkErrorTest` assertion that pins 37
so a future move off the frontier fails the suite instead of the hosts.

**Still missing on #4204, and it does not compile without it:** Swift's
`PlatformWalletError` has no `.shieldedInviteAlreadyClaimed` case, and its
`init(result:)` switches exhaustively over `PlatformWalletResultCode` with no
`default:`. Adding the raw case without the matching `init(result:)` arm makes
that switch non-exhaustive, so the Swift package fails to build at
`d78b940a03`. Rule 5's Swift clause is therefore not yet satisfied — treat the
Swift mirror as incomplete until #4204 adds the typed error case and its

> conversion arm.

Reviewer: <!-- thepastaclaw-review v1 finding=1395b65993d5 dedupe=106b52edb140376b -->
**🟡 Suggestion: Close the mirror gaps fixed on current PR heads**

New current-head drift. At #4204 head `4efecd5b71`, Swift now has the raw `errorShieldedInviteAlreadyClaimed = 37` case, generated-C `init(ffi:)` arm, typed `PlatformWalletError.shieldedInviteAlreadyClaimed` case, and exhaustive `init(result:)` arm, so this paragraph no longer describes the current branch. Likewise, #4256 head `862036b18d` mirrors codes 32 and 33 through all three Swift locations, contrary to the known-gap section at lines 440–448. Mark both gaps resolved and refresh the recorded PR-head SHAs.

<sub>source: ['codex']</sub>

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `ae21604` — _Close the mirror gaps fixed on current PR heads_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

Author: Fixed in `ae21604bae`. Both mirror gaps are closed in the registry: #4204's Swift/Kotlin/typed mirrors for code 37 and #4256's mirrors for 32/33 are recorded as resolved, and the recorded PR-head SHAs were refreshed to the post-renumber heads.

| `#4191` | **Merged into `v4.2-dev` on 2026-08-04** (`0e2282b586`). In this crate it only ever touched `dashpay.rs`; it claimed no code, so the merged table gained nothing from it — though #4183, which was stacked on it, did |

`#4186`, `#4194`, `#4195` and `#4243` are worth naming explicitly: each *does*
modify `error.rs`, but none of them adds a discriminant — `#4243`'s change maps
new wallet errors onto the **existing** `ErrorInvalidParameter`. Touching
`error.rs` is not the same as claiming an integer, and this list tracks the
latter.

**`#4277` is the merged precedent for that distinction.** It merged into
`v4.2-dev` on 2026-08-04 (`6704a41a85`) with a change to this crate's
`error.rs`, and it claimed no integer: it routes
`PlatformWalletError::TxMetadataPayloadTooLarge` onto the existing
`ErrorInvalidParameter`, with an in-line comment saying it does so deliberately
"so no new numeric code churns the Swift/Kotlin mirror enums". The merged table
is unchanged by it.

The **inherited-code table** that used to sit here is gone, and its single row is
worth recording rather than deleting:

| Code | Name | Carried by | Allocated to |
| ---: | --- | --- | --- |

> | 31 | `ErrorSigningKeyUnavailable` | #4204, #4259 | #4183 |

Reviewer: <!-- thepastaclaw-review v1 finding=ab05e07536b4 dedupe=5f918c90b14fd394 -->
**🟡 Suggestion: Keep the inherited-code table wording consistent**

New latest-delta finding. The text says the inherited-code table is gone and then immediately renders that table with its remaining row. Because the row is intentionally retained as provenance after code 31 merged, describe the table as reduced rather than removed.

```suggestion
The inherited-code table has been reduced to one row and retained for provenance:
```

<sub>source: ['coderabbit']</sub>

Author: Fixed in `695ca92ea8`. The inherited-code table is now described as reduced to one row and retained for provenance, rather than "gone", since the row is rendered immediately below.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `695ca92` — _Keep the inherited-code table wording consistent_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

| 27 | `ErrorPersisterFatal` | #3968 | Contradicts **merged ABI** — 27 is #4268's `ErrorShutdownIncomplete` (was a #4185 collision until 2026-08-02) |
| 28 | `ErrorTransactionBroadcastRejected` | #3968 | **Renumbers a shipped code** 26 → 28 — forbidden by rule 3 |

PR `#3954`'s `ErrorShutdownIncomplete = 27` used to sit in this table. It is
gone because that claim **won**: #3954 was closed and superseded by **#4268**,
which merged 27 into `v4.2-dev` on 2026-08-02. See the collision history below.

PR `#3968` is the serious one: rule 3 forbids renumbering a code that has
shipped, and `ErrorTransactionBroadcastRejected = 26` is merged ABI. Moving it
to 28 would silently reinterpret every 26 an already-compiled host returns.
PR #3968 must keep 26 where it is and take fresh integers **from the frontier
(38+)** for its two persister codes. Its 27 is now doubly wrong: 27 is merged
ABI (`ErrorShutdownIncomplete`), so rule 3 protects it too. Note that 28 is
reserved, not free — it is not available to #3968 either.

## Contested and pending

### 32 — RESOLVED: #4204 moved to 37 (first collision this file actually caught)

> Found 2026-08-03 while assembling the `v41int13` QA integration. #4204's head
> commit `b6992a5dbc` — a review round, not the original feature work — added
> `ErrorShieldedInviteAlreadyClaimed = 32` with **no row in this file**, in
> direct violation of rule 2. 32 is allocated to `ErrorTransactionBuild` (#4247,
> also carried by #4256).
> 
> Unlike every other entry in this section, this one was not a paper conflict:
> merging #4204 into an integration that already carried
> `ErrorReservationWalletMismatch = 32` produced a hard
> `error[E0081]: discriminant value 32 assigned more than once`. Resolution of
> record: **#4204 moves 32 → 37**, the frontier. `ErrorTransactionBuild` keeps 32.
> 
> The numbering was the lesser half of the defect. The code was **unmirrored on
> both hosts** — absent from Swift's `PlatformWalletResultCode` and from Kotlin's
> `fromPlatformWalletNative`. Per rule 5 that means Swift rendered it
> `.errorUnknown` (identity lost), while Kotlin fell through to `Generic(32, …)`
> — and in any tree carrying #4185's `ErrorReservationWalletMismatch = 32`,
> Kotlin actively **misclassified** "shielded invite already claimed" as
> "reservation wallet mismatch". That is the exact silently-wrong-error-on-every-host
> failure this file's preamble describes, and it landed on the shielded-invite

Reviewer: <!-- thepastaclaw-review v1 finding=1a612d53a176 dedupe=207906b3f5ebe86e -->
**🟡 Suggestion: Name ErrorTransactionBuild as the code-32 collision**

New full-PR finding, not introduced by the latest delta. The collision peer is misidentified: #4204 commit b6992a5dbc defines ErrorShieldedInviteAlreadyClaimed = 32, while #4247 defines ErrorTransactionBuild = 32. This document's own allocation history places ErrorReservationWalletMismatch at 26/28, then 30, then 36, and the inspected repository history contains no assignment of that variant to 32. The E0081 collision and any combined-tree Kotlin misclassification therefore involved TransactionBuild, not ReservationWalletMismatch. The final every-host summary is also too broad because Swift loses identity as .errorUnknown, Kotlin preserves Generic(32, …) on #4204 alone, and Kotlin misclassifies only when the competing code-32 mapping is present.

```suggestion
Unlike every other entry in this section, this one was not a paper conflict:
merging #4204 into an integration that already carried
ErrorTransactionBuild = 32 produced a hard
error[E0081]: discriminant value 32 assigned more than once. Resolution of
record: #4204 moves 32 → 37, the frontier. ErrorTransactionBuild keeps 32.

The numbering was the lesser half of the defect. The code was unmirrored on
both hosts — absent from Swift's PlatformWalletResultCode and from Kotlin's
fromPlatformWalletNative. Per rule 5, Swift rendered it .errorUnknown and lost
the identity, while Kotlin fell through to Generic(32, …) on #4204 alone. In a
tree also carrying #4247's 32 -> PlatformWallet.TransactionBuild mapping,
Kotlin actively misclassified shielded invite already claimed as transaction
build. This is the cross-host ABI failure described in the preamble.
```

<sub>source: ['codex', 'coderabbit']</sub>

Author: Fixed in `695ca92ea8`. The code-32 collision peer is corrected: #4204 defined `ErrorShieldedInviteAlreadyClaimed = 32` (b6992a5dbc) and #4247 defines `ErrorTransactionBuild = 32` (error.rs:257) — that is the E0081 pair. `ErrorReservationWalletMismatch` never held 32 (its history runs 26/28 → 30 → 36). The "silently-wrong-on-every-host" summary was also qualified: Swift loses identity as `.errorUnknown`, Kotlin on #4204 alone preserves `Generic(32, …)`, and only a tree also carrying the competing code-32 mapping misclassifies.

Reviewer: <!-- thepastaclaw-review v1 reconcile=resolved -->
Resolved in `695ca92` — _Name ErrorTransactionBuild as the code-32 collision_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

> failure this file's preamble describes, and it landed on the shielded-invite

Reviewer: The qualification is valid, and the surrounding collision identity also needs correction: code 32 belonged to ErrorTransactionBuild on #4247, not ErrorReservationWalletMismatch on #4185. The combined fix should update both the peer discriminant and the differing Swift and Kotlin outcomes.
